### ISSUE TITLE
Resource checking occasionally stops for long amounts of time after upgrade to Concourse 4

### ISSUE THREAD

**Issue by alex1x** [concourse/concourse#2476]
## Issue description

I upgraded from 3.13 to 4. I use Docker version on AWS ECS. I use RDS for the database. I kept the same DB from 3.13 when migrating, I just put restarted my contrainers with the version 4 Docker image, added some environment variables that 4 needed (like `CONCOURSE_EXTERNAL_URL`) and started using Concourse.

Some of my resources get stuck not checking saying that there is no workers:

https://cl.ly/0V2g3d0k1f24

This stays stuck for hours (as above) -- it does check correctly if I manually run fly `check-resource`, but obviously I don't want to have to manually checking all the resources.

At the same time there are workers and some pipelines and jobs work fine:

```
$ fly -t lml workers
name          containers  platform  tags  team  state    version
21d018964fe5  185         linux     none  none  running  2.1    
29394b66b3d9  156         linux     none  none  running  2.1    
2ec6a2ddd278  164         linux     none  none  running  2.1    
a81e8bd621f3  158         linux     none  none  running  2.1  
```

Some workers have started stalling (even though their container is up and running) and remaining "hidden" from Concoures after I prune them (i.e. they stall, I prune them, and then they never come back but their container is still running seemingly fine). The logs of the container don't tell me anything... at least I don't see it.

The broken container keeps looping through the following:

```
{"timestamp":"1533743717.042217970","source":"guardian","message":"guardian.api.garden-server.run.spawned","log_level":1,"data":{"handle":"243528d9-5e09-4d98-7cc5-d644eb08b30a","id":"a483884a-2e49-43da-52d1-acb6f6079a67","session":"3.1.16080","spec":{"Path":"/opt/resource/check","Dir":"","User":"root","Limits":{},"TTY":null}}}
{"timestamp":"1533743717.160425901","source":"guardian","message":"guardian.api.garden-server.run.exited","log_level":1,"data":{"handle":"243528d9-5e09-4d98-7cc5-d644eb08b30a","id":"a483884a-2e49-43da-52d1-acb6f6079a67","session":"3.1.16080","status":0}}
{"timestamp":"1533743717.200911283","source":"guardian","message":"guardian.api.garden-server.run.exited","log_level":1,"data":{"handle":"fe023109-c670-4668-555b-2df69bebb49f","id":"f365df78-6d3b-452b-73c9-1c379f487d5f","session":"3.1.16071","status":0}}
{"timestamp":"1533743717.365927458","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"10854"}}
{"timestamp":"1533743717.366630316","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"10854"}}
{"timestamp":"1533743717.382258654","source":"guardian","message":"guardian.api.garden-server.get-properties.got-properties","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","session":"3.1.16082"}}
{"timestamp":"1533743717.384599447","source":"guardian","message":"guardian.run.started","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855"}}
{"timestamp":"1533743717.384642839","source":"guardian","message":"guardian.run.exec.start","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855.2"}}
{"timestamp":"1533743717.423793316","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"10856"}}
{"timestamp":"1533743717.424370289","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"10856"}}
{"timestamp":"1533743717.439254522","source":"guardian","message":"guardian.api.garden-server.get-properties.got-properties","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","session":"3.1.16085"}}
{"timestamp":"1533743717.446022749","source":"guardian","message":"guardian.run.started","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857"}}
{"timestamp":"1533743717.446068048","source":"guardian","message":"guardian.run.exec.start","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857.2"}}
{"timestamp":"1533743717.472429514","source":"guardian","message":"guardian.run.exec.execrunner.start","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855.2.2"}}
{"timestamp":"1533743717.475610256","source":"guardian","message":"guardian.run.exec.execrunner.read-exit-fd","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855.2.2"}}
{"timestamp":"1533743717.498328924","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"10858"}}
{"timestamp":"1533743717.499100924","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"10858"}}
{"timestamp":"1533743717.511730433","source":"guardian","message":"guardian.api.garden-server.get-properties.got-properties","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","session":"3.1.16088"}}
{"timestamp":"1533743717.514087200","source":"guardian","message":"guardian.run.started","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859"}}
{"timestamp":"1533743717.514143229","source":"guardian","message":"guardian.run.exec.start","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859.2"}}
{"timestamp":"1533743717.534667253","source":"guardian","message":"guardian.run.exec.execrunner.runc-exit-status","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855.2.2","status":0}}
{"timestamp":"1533743717.534744501","source":"guardian","message":"guardian.run.exec.execrunner.done","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855.2.2"}}
{"timestamp":"1533743717.534770727","source":"guardian","message":"guardian.run.exec.finished","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855.2"}}
{"timestamp":"1533743717.534788609","source":"guardian","message":"guardian.run.finished","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","path":"/opt/resource/check","session":"10855"}}
{"timestamp":"1533743717.534809828","source":"guardian","message":"guardian.api.garden-server.run.spawned","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"1133a5e0-2370-4610-7d22-220b9e3faf83","session":"3.1.16083","spec":{"Path":"/opt/resource/check","Dir":"","User":"root","Limits":{},"TTY":null}}}
{"timestamp":"1533743717.563344717","source":"guardian","message":"guardian.run.exec.execrunner.start","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857.2.2"}}
{"timestamp":"1533743717.570003510","source":"guardian","message":"guardian.run.exec.execrunner.read-exit-fd","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857.2.2"}}
{"timestamp":"1533743717.622368813","source":"guardian","message":"guardian.run.exec.execrunner.runc-exit-status","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857.2.2","status":0}}
{"timestamp":"1533743717.622470379","source":"guardian","message":"guardian.run.exec.execrunner.done","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857.2.2"}}
{"timestamp":"1533743717.622528315","source":"guardian","message":"guardian.run.exec.finished","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857.2"}}
{"timestamp":"1533743717.622551918","source":"guardian","message":"guardian.run.finished","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","path":"/opt/resource/check","session":"10857"}}
{"timestamp":"1533743717.622577429","source":"guardian","message":"guardian.api.garden-server.run.spawned","log_level":1,"data":{"handle":"e31b63ac-c054-45d0-40fe-ecb508be45ea","id":"26c06b87-82dd-4112-6786-d925ec8acd7a","session":"3.1.16086","spec":{"Path":"/opt/resource/check","Dir":"","User":"root","Limits":{},"TTY":null}}}
{"timestamp":"1533743717.654944658","source":"guardian","message":"guardian.run.exec.execrunner.start","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859.2.2"}}
{"timestamp":"1533743717.658095837","source":"guardian","message":"guardian.run.exec.execrunner.read-exit-fd","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859.2.2"}}
{"timestamp":"1533743717.669217110","source":"guardian","message":"guardian.api.garden-server.run.exited","log_level":1,"data":{"handle":"187bbc13-6910-4b82-728d-ba60a48666a0","id":"1133a5e0-2370-4610-7d22-220b9e3faf83","session":"3.1.16083","status":0}}
{"timestamp":"1533743717.707604885","source":"guardian","message":"guardian.run.exec.execrunner.runc-exit-status","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859.2.2","status":0}}
{"timestamp":"1533743717.707718372","source":"guardian","message":"guardian.run.exec.execrunner.done","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859.2.2"}}
{"timestamp":"1533743717.707749367","source":"guardian","message":"guardian.run.exec.finished","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859.2"}}
{"timestamp":"1533743717.707766294","source":"guardian","message":"guardian.run.finished","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","path":"/opt/resource/check","session":"10859"}}
{"timestamp":"1533743717.707782984","source":"guardian","message":"guardian.api.garden-server.run.spawned","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"cbdabb28-097a-4129-72ec-dcacfecf0dc8","session":"3.1.16089","spec":{"Path":"/opt/resource/check","Dir":"","User":"root","Limits":{},"TTY":null}}}
{"timestamp":"1533743717.893727779","source":"guardian","message":"guardian.api.garden-server.run.exited","log_level":1,"data":{"handle":"c67971a1-1663-431e-5079-90748ac18252","id":"cbdabb28-097a-4129-72ec-dcacfecf0dc8","session":"3.1.16089","status":0}}
{"timestamp":"1533743717.966351032","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"10860"}}
{"timestamp":"1533743717.966938257","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"10860"}}
{"timestamp":"1533743717.990564108","source":"guardian","message":"guardian.api.garden-server.get-properties.got-properties","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","session":"3.1.16091"}}
{"timestamp":"1533743717.997989893","source":"guardian","message":"guardian.run.started","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861"}}
{"timestamp":"1533743717.998047113","source":"guardian","message":"guardian.run.exec.start","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861.2"}}
{"timestamp":"1533743718.079950094","source":"guardian","message":"guardian.run.exec.execrunner.start","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861.2.2"}}
{"timestamp":"1533743718.083584785","source":"guardian","message":"guardian.run.exec.execrunner.read-exit-fd","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861.2.2"}}
{"timestamp":"1533743718.130803347","source":"guardian","message":"guardian.run.exec.execrunner.runc-exit-status","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861.2.2","status":0}}
{"timestamp":"1533743718.130871534","source":"guardian","message":"guardian.run.exec.execrunner.done","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861.2.2"}}
{"timestamp":"1533743718.130984306","source":"guardian","message":"guardian.run.exec.finished","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861.2"}}
{"timestamp":"1533743718.130997658","source":"guardian","message":"guardian.run.finished","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","path":"/opt/resource/check","session":"10861"}}
{"timestamp":"1533743718.131008625","source":"guardian","message":"guardian.api.garden-server.run.spawned","log_level":1,"data":{"handle":"0b0217de-9b9a-48f5-41ee-85ddfa9175e2","id":"2464fd59-d3cb-494b-7096-7d3f390fd5c7","session":"3.1.16092","spec":{"Path":"/opt/resource/check","Dir":"","User":"root","Limits":{},"TTY":null}}}
{"timestamp":"1533743718.788257122","source":"guardian","message":"guardian.api.garden-server.run.exited","log_level":1,"data":{"handle":"1260dcc2-cd3a-4240-6570-ab6aa713b271","id":"98f9b2d4-eaba-4c90-6fa8-3474666aea4e","session":"3.1.16077","status":0}}
{"timestamp":"1533743718.980683804","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"10862"}}
{"timestamp":"1533743718.981309891","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"10862"}}
{"timestamp":"1533743718.993374825","source":"guardian","message":"guardian.api.garden-server.get-properties.got-properties","log_level":1,"data":{"handle":"91352a22-604b-4744-5289-5735c3b6a452","session":"3.1.16094"}}
{"timestamp":"1533743718.995661020","source":"guardian","message":"guardian.run.started","log_level":1,"data":{"handle":"91352a22-604b-4744-5289-5735c3b6a452","path":"/opt/resource/check","session":"10863"}}
{"timestamp":"1533743718.995718956","source":"guardian","message":"guardian.run.exec.start","log_level":1,"data":{"handle":"91352a22-604b-4744-5289-5735c3b6a452","id":"91352a22-604b-4744-5289-5735c3b6a452","path":"/opt/resource/check","session":"10863.2"}}
{"timestamp":"1533743719.019476891","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"10864"}}
{"timestamp":"1533743719.020248652","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"10864"}}
```

^ To confirm, this is a stalled container and it keeps producing logs like that continuously.

## Summary

After upgrade to Concourse 4

1) Check-resources fail "forever" (until checked manually)
2) Workers stall

Can you help?

**Comment by eedwards-sk** [comment]
I've been seeing similar 'hanging' waiting on resources since upgrading to 4.0.0, but haven't seen stalling workers or those log entries.

I can stand up a fresh new stack (so fresh db etc) locally and fairly regularly get into situations where I kick off a pipeline build and it just sits at the preparing stage for sometimes 2-3 minutes.

**Comment by alex1x** [comment]
I just noticed:

```
# ps aux | grep garden-init | wc -l
683
```

These look like this:

```
4294967294 128991 0.0  0.0  1020     4 ?        Ss   20:17   0:00 /tmp/garden-init
4294967294 129049 0.0  0.0  1020     4 ?        Ss   12:50   0:00 /tmp/garden-init
4294967294 129117 0.0  0.0  1020     4 ?        Ss   20:03   0:00 /tmp/garden-init
4294967294 129269 0.0  0.0  1020     4 ?        Ss   20:20   0:00 /tmp/garden-init
4294967294 129342 0.0  0.0  1020     4 ?        Ss   12:50   0:00 /tmp/garden-init
4294967294 129373 0.0  0.0  1020     4 ?        Ss   20:17   0:00 /tmp/garden-init
4294967294 129538 0.0  0.0  1020     4 ?        Ss   20:41   0:00 /tmp/garden-init
4294967294 129623 0.1  0.0  1020     4 ?        Ss   20:59   0:00 /tmp/garden-init
4294967294 130047 0.0  0.0  1020     4 ?        Ss   12:50   0:00 /tmp/garden-init
4294967294 130433 0.0  0.0  1020     4 ?        Ss   20:03   0:00 /tmp/garden-init
4294967294 130457 0.0  0.0  1020     4 ?        Ss   20:53   0:00 /tmp/garden-init
4294967294 130584 0.0  0.0  1020     4 ?        Ss   20:18   0:00 /tmp/garden-init
4294967294 130713 0.0  0.0  1020     4 ?        Ss   Aug08   0:00 /tmp/garden-init
4294967294 130742 0.0  0.0  1020     4 ?        Ss   Aug08   0:00 /tmp/garden-init
4294967294 130793 0.0  0.0  1020     4 ?        Ss   Aug08   0:00 /tmp/garden-init
```

I don't know what that user id is either.

If I run `lsof | wc -l` on my stalled worker's container (see above for more context) I see this:

```
# lsof | wc -l
[...]
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
lsof: no pwd entry for UID 4294967294
107460
```

 Can you help troubleshoot this further?

**Comment by eedwards-sk** [comment]
An update from my side, I noticed my laptop was getting very warm while the pipeline was sitting in the 'pending' state on a build. I checked `docker stats` and the worker was running very high CPU.

I checked logs and it looks like it was clearing out old volumes.

This makes perfect sense since I'm extracting some repos with thousands of files that take a while to clean up, so I think my issue is just the cleanup behavior 'working as intended' (although it'd be nice if it was surfaced in the UI somewhat, e.g. 'worker busy' or something).
